crypto/hashtable/hashtable.c

```diff
@@ -0,0 +1,682 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ *
+ *
+ *
+ * Notes On hash table design and layout
```

Comment:
Why criteria were used to accept this hash function / table?

In what sense is this _optimal_?

Reply:
I'm not sure what you mean by optimal in this case.  I experimented with several designs, and based on run time measurements of the lhash_test, comparing the same hash function and data inputs between this and lhash, I found this did about 6% better.  Now thats not a real world use case of course, but for lack of a better definitive test, it seems reasonable to me.

Reply:
How much time does lhash contribute to more normal cases?  I wonder if we're optimising something tiny.
e.g. 6% of 2% isn't worth the effort IMO.

Reply:
I haven't quantified the total amount of time spent in hash tables yet, but we have another issue here:
https://github.com/openssl/project/issues/440

In which we've found that the number of calls to namemap_name2num is called a very large number of times (in the multiple millions).  Reducing that number of calls is high on the list after this, but given that each of those calls is a lookup to a hash table, that alone suggests that an improved hash table design is worth considering.

Performance also isn't the only reason to look at this.  The ubsan issues we've faced with clang 17+ are a real problem, which require large thunking efforts in several code areas, of which LHASH is a significant part.  This implementation allows us to avoid alot of that problem.

Reply:
FWIW, with some of the hashtable updates I've been working on (using the handshake test based on the conversation [here](https://github.com/openssl/project/issues/440), I reran the tests between an old and new build (old being the head of the master branch, new being the same head, plus the new hashtable along with a conversion of the core_namemap api to use it), these are the results I currently get:
|threads|old usec|old handshakes|new usec|new handshakes| perf delta|
|-------|--------|--------------|--------|--------------|-----------|
|2|676|1463|659|1496|2.5700%
|4|1057|2783|1033|2839|2.3200%
|8|1350|5107|1318|5245|2.4200%
|16|1765|8236|1679|8699|5.1200%
|32|2585|11172|2463|11645|4.9500%
|64|5317|11537|5091|12036|4.4300%

so with the new implementation, along with updating the core_namemap code to use it, I get a throughput benefit of 2.3 to 4.9% .  I can run some of the other tests if you like, but this is the one I've been focused on most recently